### DISCUSSION TITLE
Prisma homepage database table implementation

### DISCUSSION THREAD

**Discussion by ryukikikie** [prisma/prisma#3345] [Q&A] [answered]
This is a question out of curiosity.

I've been wondering what kind of library/framework is used in https://www.prisma.io/ to display a component like below especially the database part.
<img width="1077" alt="Screen Shot 2020-08-19 at 11 27 34" src="https://user-images.githubusercontent.com/58591089/90585085-fbca9a80-e20e-11ea-91a5-2fa32acaafe3.png">

Anyone if it's fine, would like to share the tech used? I'd like to implement something similar to my personal project.

Many thanks!

**Comment by matthewmueller** [comment, marked as answer]
I wish I could say there's a simple tool that will handle this easily for you. The truth is, there isn't – it's just a lot of work. 

The way to think about this animation is there's a bunch of frames (states at a point in time). Each of these frames lasts only for a moment, before switching to the next frame.

In React, that would look something like this:

```js
let i = 0
const frames = [ ... ]

function nextFrame(frame) {
  this.setState({ frame })  
  setTimeout(function() {
    nextFrame(frames[i++])
  }, frame.duration)
}

nextFrame(frames[i])
```

This is an example of 1 frame in our animation:

```js
{
  duration: between(200, 250),
  table: [
    [false, 'id', 'firstName', 'email', 'active'],
    [false, '1', 'Patrick', 'patrick@prisma.io', 'true'],
    [false, '2', 'Eden', 'eden@prisma.io', 'true'],
    [false, '3', 'Dylan', 'dylan@prisma.io', 'false'],
    [false, '4', 'Anki', 'anki@prisma.io', 'true'],
    [false, '5', 'Juraci', 'juraci@prisma.io', 'false'],
    [false, '6', 'Alice', 'alice@prisma.io', 'true'],
  ],
  lines: [
    [['comment', '// Finding records']],
    [
      ['keyword', 'await'],
      ['punctuation', ' '],
      ['keyword', 'prisma'],
      ['punctuation', '.'],
      ['keyword', 'users'],
      ['punctuation', '.'],
      ['keyword', 'findMany'],
      ['punctuation', '({'],
    ],
    [
      ['punctuation', '  '],
      ['property', 'where'],
      ['punctuation', ': {'],
    ],
    [
      ['punctuation', '    '],
      ['property', 'active'],
      ['punctuation', ': '],
      ['boolean', 't'],
      ['cursor', 'solid'],
    ],
    [
      ['punctuation', '  '],
      ['punctuation', '}'],
    ],
    [['punctuation', '})']],
  ],
  suggests: [
    [true, [['highlight', 't'], 'rue'], 'boolean'],
  ],
}
```

In this animation, there's 67 individual frames. Switching between frames quickly is what drives the animation. 
 Lastly, we wrap a couple divs with [framer-motion](https://github.com/framer/motion) at the end to give it a springy feel.

Good luck!

**Reply by ryukikikie** [reply]
Thank you very much Matthew for the detailed explanation and sharing the implementation!
One last question, for the database's look(Really cool combination of black and green), did you use any framework to make it look that way or multiple of simple html elements?

**Reply by matthewmueller** [reply]
The look and feel is @kuldar magic. The implementation is just a `<table>`.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
